src/Bridge/Doctrine/Common/Filter/BooleanFilterTrait.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Doctrine\Common\Filter;
+
+use ApiPlatform\Core\Exception\InvalidArgumentException;
+use Doctrine\DBAL\Types\Type as DBALType;
+
+/**
+ * Trait for filtering the collection by boolean values.
+ *
+ * Filters collection on equality of boolean properties. The value is specified
+ * as one of ( "true" | "false" | "1" | "0" ) in the query.
+ *
+ * For each property passed, if the resource does not have such property or if
+ * the value is not one of ( "true" | "false" | "1" | "0" ) the property is ignored.
+ *
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ * @author Teoh Han Hui <teohhanhui@gmail.com>
+ */
+trait BooleanFilterTrait
```

Comment:
Why not an internal abstract class? It would prevent to introduce a bunch of new interfaces.

Reply:
It is not possible. The code is shared between the `Orm` and the `Odm` boolean filter. They need to extend their own `AbstractContextAwareFilter`.

Reply:
Also, which interfaces are you talking about?

Reply:
What about something like the following?

```php
/** @internal */
abstract class AbstractBooleanFilter
{
    // ...
}

class OrmBooleanFilter extends AbstractBooleanFilter implements OrmAbstractContextAwareFilter
{
    // ...
}

final class MongoBooleanFilter extends AbstractBooleanFilter implements MongoAbstractContextAwareFilter
{
    // ...
}
```

Doing so, interfaces like `SearchFilterInterface` can be removed.

Reply:
`OrmAbstractContextAwareFilter` and `MongoAbstractContextAwareFilter` are abstract classes not interfaces. You seem to suggest the first way I tried before but which needed to use only one interface for filters.

Reply:
I'm not sure to understand. You may have everything in the abstract class and keep type hints:

```php
/** @internal */
abstract class AbstractBooleanFilter
{
    protected function doSomething($noTypeHint) {
        // contains the actual logic, shared
    }
}

final class MongoBooleanFilter extends AbstractBooleanFilter implements MongoAbstractContextAwareFilter
{
    public function something(TypeHint $typeHinted) {
        // preserves the typehint, delegate to the shared method
        return $this->doSomething($typeHinted);
    }
}
```

WDYT?